test/js_test.go

```diff
@@ -955,6 +955,381 @@ func TestJetStreamSubscribe(t *testing.T) {
 	}
 }
 
+func TestPullSubscribeFetchBatch(t *testing.T) {
+	s := RunBasicJetStreamServer()
+	defer shutdownJSServerAndRemoveStorage(t, s)
+
+	nc, js := jsClient(t, s)
+	defer nc.Close()
+
+	_, err := js.AddStream(&nats.StreamConfig{
+		Name:     "TEST",
+		Subjects: []string{"foo"},
+	})
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+
+	t.Run("basic fetch", func(t *testing.T) {
+		defer js.PurgeStream("TEST")
+		sub, err := js.PullSubscribe("foo", "")
+		if err != nil {
+			t.Fatalf("Unexpected error: %s", err)
+		}
+		for i := 0; i < 5; i++ {
+			if _, err := js.Publish("foo", []byte("msg")); err != nil {
+				t.Fatalf("Unexpected error: %s", err)
+			}
+		}
+		res, err := sub.FetchBatch(10)
+		if err != nil {
+			t.Fatalf("Unexpected error: %s", err)
+		}
+		go func() {
+			time.Sleep(10 * time.Millisecond)
+			for i := 0; i < 5; i++ {
+				js.Publish("foo", []byte("msg"))
+			}
+		}()
+		msgs := make([]*nats.Msg, 0)
+		for msg := range res.Messages() {
+			msgs = append(msgs, msg)
+		}
+		if res.Error() != nil {
+			t.Fatalf("Unexpected error: %s", res.Error())
+		}
+		if len(msgs) != 10 {
+			t.Fatalf("Expected %d messages; got: %d", 10, len(msgs))
+		}
+	})
+
+	t.Run("multiple concurrent fetches", func(t *testing.T) {
+		defer js.PurgeStream("TEST")
+		sub, err := js.PullSubscribe("foo", "")
+		if err != nil {
+			t.Fatalf("Unexpected error: %s", err)
+		}
+		for i := 0; i < 50; i++ {
+			if _, err := js.Publish("foo", []byte("msg")); err != nil {
+				t.Fatalf("Unexpected error: %s", err)
+			}
+		}
+		var r1, r2, r3 nats.MessageBatch
+		started := &sync.WaitGroup{}
+		started.Add(3)
+		errs := make(chan error, 3)
+		go func() {
+			r1, err = sub.FetchBatch(10)
+			if err != nil {
+				errs <- err
+			}
+			started.Done()
+		}()
+		go func() {
+			r2, err = sub.FetchBatch(10)
+			if err != nil {
+				errs <- err
+			}
+			started.Done()
+		}()
+		go func() {
+			r3, err = sub.FetchBatch(10)
+			if err != nil {
+				errs <- err
+			}
+			started.Done()
+		}()
+		// wait until batch results are available
+		started.Wait()
+
+		// check if any FetchBatch call returned an error
+		select {
+		case err := <-errs:
+			t.Fatalf("Error initializing fetch: %s", err)
+		default:
+		}
+
+		var msgsReceived int
+		for msgsReceived < 30 {
+			select {
+			case <-r1.Messages():
+				msgsReceived++
+			case <-r2.Messages():
+				msgsReceived++
+			case <-r3.Messages():
+				msgsReceived++
+			case <-time.After(1 * time.Second):
+				t.Fatalf("Timeout waiting for incoming messages")
+			}
+		}
+		if r1.Error() != nil {
+			t.Fatalf("Unexpected error: %s", r1.Error())
+		}
+		if r2.Error() != nil {
+			t.Fatalf("Unexpected error: %s", r2.Error())
+		}
+		if r3.Error() != nil {
+			t.Fatalf("Unexpected error: %s", r3.Error())
+		}
+		if !r1.Done() {
```

Comment:
this one seems to be a data race... probably due to marking `bm.done` bool in a different goroutine without locks and then checking it here, maybe it was needed the `Done() <-chan struct{}` approach after all to avoid locks... I'm also good on punting adding `Done()` to the `MessageBatch` interface for now if it becomes problematic.

Reply:
Yes, my bad. I revise my statement about not liking `Done() <-chan struct{}` 😅 It does not force anyone to use `select` + it can always be used just like a boolean with `len(res.Done()) == 0`. I changed that and fixed the above data race as well.